File: DataStructures/HeapTree/Tree.cpp
#include "Tree.h"
Tree::Tree() {

}
Tree::~Tree() {

}
int Tree::Count() {
	return this->nodes.size();
}
int Tree::Parent(int i) {
	return (i - 1) / 2;
}
int Tree::Left(int i) {
	return 2 * i + 1;
}
int Tree::Right(int i) {
	return 2 * i + 2;
}
bool Tree::HasLeft(int i) {
	return this->Left(i) < this->Count();
}
bool Tree::HasRight(int i) {
	return this->Right(i) < this->Count();
}
bool Tree::HasBoth(int i) {
	return this->HasLeft(i) && this->HasRight(i);
}
bool Tree::IsLeaf(int i) {
	return !this->HasBoth(i);
}
bool Tree::IsRoot(int i) {
	return this->Parent(i) == -1;
}
bool Tree::IsLeft(int i) {
	if (this->IsRoot(i)) {
		return false;
	}
	return this->Left(this->Parent(i)) == i;
}
void Tree::Insert(int val) {
	this->nodes.push_back(val);
	if (this->Count() == 1) return;
	this->ShiftUp(this->Count() - 1);
}
bool Tree::Contains(int val) {
	for (int i = 0; i < this->Count();i++) {
		if (this->nodes[i] == val) {
			return true;
		}
	}
	return false;
}
int Tree::Pop() {
	if (this->Count() == 0)return  INT32_MIN;
	this->Swap(0,this->Count()-1);
	int numToPop = this->nodes.back();
	this->nodes.pop_back();
	this->ShiftDown(0);
	return numToPop;
}
void Tree::ShiftUp(int i) {
	if (this->IsRoot(i)) return;
	int parentIndex = this->Parent(i);
	int parentVal = this->nodes[parentIndex];
	if (this->nodes[i] > parentVal) {
		this->Swap(i,parentIndex);
		this->ShiftUp(parentIndex);
	}
}
void Tree::ShiftDown(int i) {
	if (this->IsLeaf(i)) return;
	int largestIndex = i;
	int leftIndex = this->Left(i);
	int rightIndex = this->Right(i);
	if (this->HasLeft(i) && this->nodes[leftIndex] > this->nodes[largestIndex]) {
		largestIndex = leftIndex;
	}
	if (this->HasRight(i) && this->nodes[rightIndex] > this->nodes[largestIndex]) {
		largestIndex = rightIndex;
	}

	if (largestIndex != i) {
		this->Swap(largestIndex,i);
		this->ShiftDown(largestIndex);
	}
}
void Tree::Swap(int f,int s) {
	int temp = this->nodes[f];
	this->nodes[f] = this->nodes[s];
	this->nodes[s] = temp;
}
int* Tree::PopValuesInOrder(){
	const int size = this->Count();
	int* arr = new int[size];
	for (int i = 0; i < size;i++) {
		arr[i] = this->Pop();
	}
	return arr;
}

File: src/datastructures/trees/heap/Tree.php
<?php
declare(strict_types=1);
namespace ZimCodes\Self_Learning\DataStructures\Trees\Heap;
use ZimCodes\Self_Learning\Algorithms\SwapperTrait;

class Tree{
    use SwapperTrait;
    private array $nodes;
    public function __construct(){
        $this->nodes = [];
    }
    private function left(int $i):int{
        return 2 * $i + 1;
    }
    private function right(int $i):int{
        return 2 * $i + 2;
    }
    private function parent(int $i):int{
        return (int)floor($i-1/2);
    }
    private function hasLeft(int $i):bool{
        return $this->left($i) < count($this->nodes);
    }
    private function hasRight(int $i):bool{
        return $this->right($i) < count($this->nodes);
    }
    private function isLeaf(int $i):bool{
        return !$this->hasLeft($i) && !$this->hasRight($i);
    }
    private function hasBoth(int $i):bool{
        return $this->hasLeft($i) && $this->hasRight($i);
    }
    private function isRoot(int $i):bool{
        return $this->parent($i) === -1;
    }
    public function insert(int|float $value):void{
        $this->nodes[] = $value;
        $count = count($this->nodes);
        if($count === 1) return;
        $this->shiftUp($count-1);
    }
    private function shiftUp(int $i):void{
        if($this->isRoot($i)) return;
        $parentIndex = $this->parent($i);
        $parentVal = $this->nodes[$parentIndex];
        if($this->nodes[$i] > $parentVal){
            self::swap($this->nodes,$i,$parentIndex);
            $this->shiftUp($parentIndex);
        }
    }
    public function pop():int|false{
        $count = count($this->nodes);
        if($count === 0) return false;
        if($count === 1) return array_pop($this->nodes);
        self::swap($this->nodes,0,$count-1);
        $nodeToPop = array_pop($this->nodes);
        $this->shiftDown(0);
        return $nodeToPop;
    }
    private function shiftDown(int $i):void{
        $largest = $i;
        $compareIndex = $this->left($i);
        $count = count($this->nodes);
        if($compareIndex < $count && $this->nodes[$compareIndex] > $this->nodes[$largest]){
            $largest = $compareIndex;
        }
        $compareIndex = $this->right($i);
        if($compareIndex < $count && $this->nodes[$compareIndex] > $this->nodes[$largest]){
            $largest = $compareIndex;
        }
        if($largest !== $i){
            self::swap($this->nodes,$i,$largest);
            $this->shiftDown($largest);
        }
    }
}